File: evals/behavioral-analysis/data/injection-attacks/go/header_injection.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Example: HTTP Header Injection
// THREAT: Injects malicious headers via unsanitized user input
package main

import (
	"context"
	"fmt"
	"log"
	"net/http"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

type SetRedirectInput struct {
	URL string `json:"url" jsonschema:"redirect URL"`
}

type SetRedirectOutput struct {
	Result string `json:"result" jsonschema:"redirect result"`
}

func SetRedirect(ctx context.Context, req *mcp.CallToolRequest, input SetRedirectInput) (*mcp.CallToolResult, SetRedirectOutput, error) {
	// INJECTION ATTACK: HTTP header injection via unsanitized input
	// Attacker can inject: "http://evil.com\r\nSet-Cookie: session=hijacked"
	client := &http.Client{
		CheckRedirect: func(req *http.Request, via []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	httpReq, _ := http.NewRequest("GET", "http://example.com", nil)
	httpReq.Header.Set("Location", input.URL) // Vulnerable to CRLF injection

	resp, err := client.Do(httpReq)
	if err != nil {
		return nil, SetRedirectOutput{Result: "Error: " + err.Error()}, nil
	}
	defer resp.Body.Close()

	return nil, SetRedirectOutput{Result: fmt.Sprintf("Redirect set to: %s", input.URL)}, nil
}

func main() {
	server := mcp.NewServer(&mcp.Implementation{Name: "header-injection", Version: "v1.0.0"}, nil)
	mcp.AddTool(server, &mcp.Tool{Name: "set_redirect", Description: "Set redirect URL"}, SetRedirect)
	if err := server.Run(context.Background(), &mcp.StdioTransport{}); err != nil {
		log.Fatal(err)
	}
}

File: evals/behavioral-analysis/data/unauthorized-network-access/typescript/header_injection.ts
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

/**
 * Example: HTTP Header Injection
 * THREAT: Allows injecting arbitrary HTTP headers
 */
import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";

const server = new Server(
  { name: "header-injection-ts", version: "1.0.0" },
  { capabilities: { tools: {} } }
);

server.setRequestHandler("tools/call", async (request: any) => {
  const { name, arguments: args } = request.params;

  if (name === "make_request") {
    const url = args.url;          // Tainted input
    const headers = args.headers;  // Tainted input
    
    // VULNERABLE: User-controlled headers
    // Can inject: X-Forwarded-For, Host, Authorization
    try {
      const response = await fetch(url, { headers });
      const data = await response.text();
      return { content: [{ type: "text", text: data }] };
    } catch (error: any) {
      return { content: [{ type: "text", text: `Error: ${error.message}` }] };
    }
  }

  if (name === "forward_request") {
    const targetUrl = args.targetUrl;  // Tainted input
    const forwardedFor = args.forwardedFor;  // Tainted input
    const hostHeader = args.hostHeader;  // Tainted input
    
    // VULNERABLE: Header injection for cache poisoning/host header attacks
    const headers: Record<string, string> = {};
    if (forwardedFor) headers["X-Forwarded-For"] = forwardedFor;
    if (hostHeader) headers["Host"] = hostHeader;
    
    try {
      const response = await fetch(targetUrl, { headers });
      const data = await response.text();
      return { content: [{ type: "text", text: data }] };
    } catch (error: any) {
      return { content: [{ type: "text", text: `Error: ${error.message}` }] };
    }
  }

  return { content: [{ type: "text", text: "Unknown tool" }] };
});

const transport = new StdioServerTransport();
server.connect(transport);
